### PR TITLE
kernel: Update threading/synch

### PR DESCRIPTION
**Pull request by bunnei**
Various updates and improvements to our kernel threading and synchronization

### CODE REVIEW

File: src/core/hle/kernel/k_light_condition_variable.cpp [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+#include "core/hle/kernel/k_scoped_scheduler_lock_and_sleep.h"
+#include "core/hle/kernel/k_thread_queue.h"
+#include "core/hle/kernel/svc_results.h"
+
+namespace Kernel {
+
+namespace {
+
+class ThreadQueueImplForKLightConditionVariable final : public KThreadQueue {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_light_condition_variable.cpp:16]: Our convention is to put the private bits after the public interface

File: src/core/hle/kernel/k_address_arbiter.cpp [hunk 2]
```diff
@@ -85,6 +86,27 @@ bool UpdateIfEqual(Core::System& system, s32* out, VAddr address, s32 value, s32
     return true;
 }
 
+class ThreadQueueImplForKAddressArbiter final : public KThreadQueue {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_address_arbiter.cpp:90]: Ditto

File: src/core/hle/kernel/k_synchronization_object.cpp [hunk 3]
```diff
@@ -8,11 +8,70 @@
...
+#include "core/hle/kernel/k_thread_queue.h"
 #include "core/hle/kernel/kernel.h"
 #include "core/hle/kernel/svc_results.h"
 
 namespace Kernel {
 
+namespace {
+
+class ThreadQueueImplForKSynchronizationObjectWait final : public KThreadQueueWithoutEndWait {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_synchronization_object.cpp:20]: Ditto

File: src/core/hle/kernel/k_thread.cpp [hunk 4]
```diff
@@ -56,6 +56,35 @@ static void ResetThreadContext64(Core::ARM_Interface::ThreadContext64& context,
...
+namespace {
+
+class ThreadQueueImplForKThreadSleep final : public KThreadQueueWithoutEndWait {
+public:
+    explicit ThreadQueueImplForKThreadSleep(KernelCore& kernel_)
+        : KThreadQueueWithoutEndWait(kernel_) {}
+};
+
+class ThreadQueueImplForKThreadSetProperty final : public KThreadQueue {
+private:
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread.cpp:68]: Ditto

File: src/core/hle/kernel/k_thread.cpp [hunk 5]
```diff
@@ -56,6 +56,35 @@ static void ResetThreadContext64(Core::ARM_Interface::ThreadContext64& context,
...
+    KThread::WaiterList* m_wait_list;
+
+public:
+    explicit ThreadQueueImplForKThreadSetProperty(KernelCore& kernel_, KThread::WaiterList* wl)
+        : KThreadQueue(kernel_), m_wait_list(wl) { // ...
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread.cpp:73]:
```suggestion
        : KThreadQueue(kernel_), m_wait_list(wl) {}
```

File: src/core/hle/kernel/k_synchronization_object.cpp [hunk 6]
```diff
@@ -8,11 +8,70 @@
...
+
+private:
+    KSynchronizationObject** m_objects;
+    ThreadListNode* m_nodes;
+    s32 m_count;
+
+public:
+    ThreadQueueImplForKSynchronizationObjectWait(KernelCore& kernel_, KSynchronizationObject** o,
+                                                 ThreadListNode* n, s32 c)
+        : KThreadQueueWithoutEndWait(kernel_), m_objects(o), m_nodes(n), m_count(c) { // ...
```
Comment (lioncash) [inline src/core/hle/kernel/k_synchronization_object.cpp:31]:
```suggestion
        : KThreadQueueWithoutEndWait(kernel_), m_objects(o), m_nodes(n), m_count(c) {}
```

File: src/core/hle/kernel/k_thread.h [hunk 7]
```diff
@@ -197,13 +195,19 @@ class KThread final : public KAutoObjectWithSlabHeapAndContainer<KThread, KSynch
...
+        return synced_index;
+    }
+
+    constexpr void SetWaitResult(ResultCode wait_res) {
         wait_result = wait_res;
     }
 
-    [[nodiscard]] ResultCode GetWaitResult(KSynchronizationObject** out) const {
-        *out = synced_object;
+    constexpr ResultCode GetWaitResult() const {
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread.h:210]: This should still be `[[nodiscard]]` shouldn't it?

File: src/core/hle/kernel/k_thread_queue.h [hunk 8]
```diff
@@ -4,78 +4,32 @@
...
 
+#include "core/hle/kernel/k_scheduler.h"
 #include "core/hle/kernel/k_thread.h"
 
 namespace Kernel {
 
 class KThreadQueue {
 public:
     explicit KThreadQueue(KernelCore& kernel_) : kernel{kernel_} {}
+    virtual ~KThreadQueue(){};
```
Comment (lioncash) [inline src/core/hle/kernel/k_thread_queue.h:15]:
```suggestion
    virtual ~KThreadQueue() = default;
```

**Comment by FernandoS27** [comment]
I shall not mark until A. you fix the linux compile errors and B. you play Pokemon Sword start to finish with it.

**Comment by bunnei** [comment]
Superseded!

**Pull request closed without merging**